lib/kube/proxy/streamproto/proto.go

```diff
@@ -0,0 +1,280 @@
+	conn *websocket.Conn
+
+	// A stream of incoming session packets.
+	in chan []byte
+
+	// Optionally contains a partially read session packet.
+	currentIn []byte
+
+	// A list of resize requests.
+	resizeQueue chan *remotecommand.TerminalSize
+
+	// A notification channel for force termination requests.
+	forceTerminate chan struct{}
+
+	writeSync   sync.Mutex
+	done        chan struct{}
+	closedC     sync.Once
+	closed      bool
+	MFARequired bool
+	Mode        types.SessionParticipantMode
+}
+
+// NewSessionStream creates a new session stream.
+// The type of the handshake parameter determines if this is the client or server end.
+func NewSessionStream(conn *websocket.Conn, handshake interface{}) (*SessionStream, error) {
+	s := &SessionStream{
+		conn:           conn,
+		in:             make(chan []byte),
+		done:           make(chan struct{}),
+		resizeQueue:    make(chan *remotecommand.TerminalSize, 1),
+		forceTerminate: make(chan struct{}),
+	}
+
+	clientHandshake, isClient := handshake.(ClientHandshake)
+	serverHandshake, ok := handshake.(ServerHandshake)
+
+	if !isClient && !ok {
+		return nil, trace.BadParameter("Handshake must be either client or server handshake, got %T", handshake)
+	}
+
+	if isClient {
+		ty, data, err := conn.ReadMessage()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if ty != websocket.TextMessage {
+			return nil, trace.Errorf("Expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ServerHandshake == nil {
+			return nil, trace.Errorf("Expected websocket server handshake, got %v", msg)
+		}
+
+		s.MFARequired = msg.ServerHandshake.MFARequired
+		handshakeMsg := metaMessage{ClientHandshake: &clientHandshake}
+		dataClientHandshake, err := utils.FastMarshal(handshakeMsg)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := conn.WriteMessage(websocket.TextMessage, dataClientHandshake); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else {
+		handshakeMsg := metaMessage{ServerHandshake: &serverHandshake}
+		dataServerHandshake, err := utils.FastMarshal(handshakeMsg)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := conn.WriteMessage(websocket.TextMessage, dataServerHandshake); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		ty, data, err := conn.ReadMessage()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if ty != websocket.TextMessage {
+			return nil, trace.Errorf("Expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ClientHandshake == nil {
+			return nil, trace.Errorf("Expected websocket client handshake")
+		}
+
+		s.Mode = msg.ClientHandshake.Mode
+	}
+
+	go s.readTask()
+	return s, nil
+}
+
+func (s *SessionStream) readTask() {
```

Comment:
I'm not convinced that the goroutine running this method won't leak and hang forever.

- should you pass a context through?
- does `ReadMessage()` block? if so, how can we avoid hanging forever if a message never comes?
- not sure what the terminated flag is for, but it's initialized as false at the beginning of every iteration and never read from after it's set to true, so it ends up having no effect.

Reply:
1. I'm not sure how a context would help much here unfortunately. We could spin off a seperate goroutine to read messages but I don't know what this would improve.
2. `ReadMessage()` does block, it does however return an error when the underlying connection is closed i.e `io.EOF`. I've gone ahead and added a warn level log for non io.EOF errors.